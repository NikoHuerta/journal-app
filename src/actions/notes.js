import { db } from "../firebase/firebaseConfig";
import { collection, addDoc, updateDoc, doc , deleteDoc} from "firebase/firestore";
import { types } from "../types/types";
import { loadNotes } from "../helpers/loadNotes";
import Swal from "sweetalert2";
import { fileUpload } from "../helpers/fileUpload";


export const startNewNote = () => { //tarea async
    return async ( dispatch, getState ) => {
        const { uid } = getState().auth;

        
        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime()
        };

        try{
            const doc = await addDoc(collection(db, `${uid}/journal/notes`), newNote);
            dispatch(activeNote(doc.id, newNote));
            dispatch(addNewNote(doc.id, newNote));
        }catch(e){
            console.log(e);
        }
    }
}

export const startLoadingNotes = (uid) => { //tarea async
    return async (dispatch) => {
        const notes = await loadNotes(uid);
        dispatch(setNotes(notes));
    }
}

export const addNewNote = (id, note) => ({
    type: types.notesAddNewNote,
    payload: { 
            id,
            ...note
    }
});

export const activeNote = (id, note) => ({ //tarea sync
    type: types.notesActive,
    payload: {
        id,
        ...note
    }
});

export const setNotes = (notes) => ({ //tarea sync (local)
    type: types.notesLoad,
    payload: notes
});

export const startSaveNote = ( note ) => { //tarea async (remota)
    return async (dispatch, getState) => {
        const { uid } = getState().auth;
        // //obtener el estado actual de las notas desde state redux
        // const { notes:notesReducer } = getState().notes;

        if(!note.url){
            delete note.url;
        }

        const noteToFirestore = { ...note };
        delete noteToFirestore.id;

        const noteRef = doc(db, `${ uid }/journal/notes/${ note.id }`);
        try {
                await updateDoc(noteRef, noteToFirestore);
                dispatch(refreshNote(note.id, noteToFirestore));

                // //aÃ±adir note nueva al state
                // notesReducer.push(note);
                // dispatch(setNotes(notesReducer));

                Swal.fire('Saved', note.title, 'success');
        }catch(e){
            Swal.fire('Error Saving', e, 'error');
        }
    }
};

export const refreshNote = (id, note) => ({ //tarea sync (local)
    type: types.notesUpdated,
    payload: { 
        id,
        note: {
            id,
            ...note
        }
    }   
});

export const startUploading = (file) => {
    return async (dispatch, getState) => {
        const { active:activeNote } = getState().notes;

        Swal.fire({
                title: 'Uploading...', 
                text: 'Please wait...',
                allowOutsideClick: false,
                didOpen: () => {
                    Swal.showLoading();
                }
            });

        const fileUrl = await fileUpload(file);
        Swal.close();

        activeNote.url = fileUrl;
        dispatch(startSaveNote(activeNote));
        // console.log(fileUrl);
    }
};

export const startDeleting = (id) => {
    return async (dispatch, getState) => {
        const { uid } = getState().auth;
        const noteRef = doc(db, `${ uid }/journal/notes/${ id }`);

        try {
            await deleteDoc(noteRef);
            dispatch(deleteNote(id));

        } catch (error) {
            console.log(error);
        }
    }
};

export const deleteNote = (id) => ({
    type: types.notesDelete,
    payload: id
});

export const logoutNote = () => ({
    type: types.notesLogoutCleaning
})