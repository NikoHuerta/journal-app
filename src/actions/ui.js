import { types } from "../types/types";

export const setError = (error) => ({
    type: types.uiSetError,
    payload: error
});

export const removeError = () => ({
    type: types.uiRemoveError
});