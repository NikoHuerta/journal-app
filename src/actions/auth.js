import Swal from 'sweetalert2';

import { getAuth, signInWithPopup, createUserWithEmailAndPassword, updateProfile, signInWithEmailAndPassword, signOut } from 'firebase/auth';
import { googleAuthProvider } from "../firebase/firebaseConfig";
import { types } from "../types/types"
import { finishLoading, startLoading } from './ui';
import { logoutNote } from './notes';


export const startLoginEmailPassword = (email, password) => {
    return (dispatch) => {
        
        dispatch(startLoading());

        const auth = getAuth();
        // console.log(auth);

        return signInWithEmailAndPassword(auth, email, password)
            .then( ({user}) => {
                dispatch(login(user.uid, user.displayName));
                dispatch(finishLoading());
            })
            .catch(e => {
                dispatch(finishLoading());
                //console.log(e);
                let msg_err;
                ((e.code ==='auth/user-not-found') || (e.code ==='auth/wrong-password')) ? msg_err= 'You have entered an invalid username or password, or your account does not exist' : msg_err = 'Service error';
                Swal.fire('Error', msg_err, 'error');

            });
    }
}

export const startRegisterWithEmailPasswordName = (email, password, name) => {
    return (dispatch) => {
        const auth = getAuth();
        return createUserWithEmailAndPassword(auth, email, password)
            .then( async ({ user }) => {

                await updateProfile(user, { displayName: name });                
                dispatch(login(user.uid, user.displayName));
            })
            .catch(e => {
                
                let msg_err;
                (e.code ==='auth/email-already-in-use') ? msg_err= 'The email address is already in use by another account.' : msg_err = 'Service internal error.';
                Swal.fire('Error', msg_err, 'error');
                return msg_err;
            });

    }
}

export const startGoogleLogin = () => {

    return(dispatch) => {
        dispatch(startLoading());

        const auth = getAuth();
        return signInWithPopup(auth, googleAuthProvider)
            .then( ({ user }) => {
                dispatch(
                    login(user.uid, user.displayName)
                );
                dispatch(finishLoading());
            })
            .catch(e => {
                Swal.fire('Error', 'The Google Login PopUp was closed.', 'error');
                dispatch(finishLoading());
                return e.code;
            });
    }
}

export const startLogout = () => {
    return (dispatch) => {
        const auth = getAuth();
        //console.log(auth);
        return signOut(auth)
            .then( ( ) => {
                dispatch(logout());
                dispatch(logoutNote());
            })
            .catch( e => {
                console.log(e);
            });
    }
}

export const login = (uid, displayName) => ({
        type: types.login,
        payload: {
            uid,
            displayName
        }
});

export const logout = () => ({
    type: types.logout
});


