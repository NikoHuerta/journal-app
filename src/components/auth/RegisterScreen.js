import React from 'react';
import { Link } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import validator from 'validator';
// import Swal from 'sweetalert2';

import { useForm } from '../../hooks/useForm';
import { removeError, setError } from '../../actions/ui';
import { startRegisterWithEmailPasswordName } from '../../actions/auth';



export const RegisterScreen = () => {

    const dispatch = useDispatch();
    const { msgError } = useSelector( state => state.ui );
    

    const [formValues, handleInputChange] = useForm({ 
        name: 'Nicolas Test',
        email: 'nicolas@huerta.cl',
        password: '123456',
        password2: '123456' 
    });

    const { name, email, password, password2 } = formValues;

    const handleRegister = (e) => {
        e.preventDefault();
        if(isFormValid()){
            // console.log('Formulario Correcto');
            dispatch( startRegisterWithEmailPasswordName(email, password, name) );
        }
    };

    // const handleError = (msg) => {
    //     Swal.fire('Error',  msg , 'warning');
    // };

    const isFormValid = () => {

        if(name.trim().length === 0){
            dispatch(setError('Name is required'));
            return false;
        } else if (!validator.isEmail(email)){
            dispatch(setError('Email is not valid'));
            return false;
        } else if (password !== password2 || password.length < 6){
            dispatch(setError('Password shoud be at least 6 characters and match each other'));
            return false;
        }
        
        dispatch(removeError());
        return true;
    }

    
    return (
        <>
            <h3 className="auth__title">Register</h3>

            <form onSubmit={ handleRegister } >

                {
                    msgError && //handleError(msgError)
                    (<div className="auth__alert-error">
                        { msgError }
                    </div>)
                    
                }


                <input 
                    type="text"
                    placeholder="Name"
                    name="name"
                    className="auth__input"
                    autoComplete="off"

                    onChange={ handleInputChange }
                    value= { name }
                />

                <input 
                    type="text"
                    placeholder="Email"
                    name="email"
                    className="auth__input"
                    autoComplete="off"

                    onChange={ handleInputChange }
                    value= { email }
                />

                <input 
                    type="password"
                    placeholder="Password"
                    name="password"
                    className="auth__input"

                    onChange={ handleInputChange }
                    value= { password }
                />
                

                <input 
                    type="password"
                    placeholder="Confirm Password"
                    name="password2"
                    className="auth__input"

                    onChange={ handleInputChange }
                    value= { password2 }
                />

                <button 
                    type="submit"
                    className="btn btn-primary btn-block mb-5"
                    // disabled= { true }
                >Register
                </button>
                
                <Link to="/auth/login" className="link">
                    Already Registered?
                </Link>
            </form>
        </>
    )
}
