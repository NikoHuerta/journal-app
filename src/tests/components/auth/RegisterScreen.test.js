import { mount } from "enzyme";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";

import configureStore from 'redux-mock-store';
import thunk from "redux-thunk";
import { startRegisterWithEmailPasswordName } from "../../../actions/auth";

import { RegisterScreen } from "../../../components/auth/RegisterScreen";
import { types } from "../../../types/types";

// jest.mock('../../../actions/auth', ()=>({
//     startRegisterWithEmailPasswordName: jest.fn()
// }));

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const initState = {
    auth: {  },
    ui: {
        loading: false,
        msgError: null
    },
    notes:{ 
        notes: [],
        active: null
     }
};
let store = mockStore(initState);
//store.dispatch = jest.fn();


const wrapper = mount(
    <Provider store={ store }>
        <MemoryRouter>
            <RegisterScreen />
        </MemoryRouter>
    </Provider>
);

describe('Pruebas en <RegisterScreen />', () => {
   
    // beforeEach(()=>{
    //     store = mockStore(initState);
    //     jest.clearAllMocks();
    // });

    test('Debe de mostrarse correctamente', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('Debe de disparar la accion de startRegisterWithEmailPasswordName', ()=>{

        const emailField = wrapper.find('input[name="email"]');
        emailField.simulate('change', {
            target:{
                value: '',
                name: 'email'
            }
        });

        wrapper.find('form').simulate('submit', {
            preventDefault(){}
        });

        const actions = store.getActions();
        //console.log(actions);
        expect(actions[0]).toEqual({
            type: types.uiSetError,
            payload: 'Email is not valid'
        });
    });

    test('Debe de mostrar la caja de alerta con el error', () => {
        const initState = {
            auth: {  },
            ui: {
                loading: false,
                msgError: 'Email is not valid'
            },
            notes:{ 
                notes: [],
                active: null
             }
        };
        const store = mockStore(initState);

        const wrapper = mount(
            <Provider store={ store }>
                <MemoryRouter>
                    <RegisterScreen />
                </MemoryRouter>
            </Provider>
        );

        expect(wrapper.find('.auth__alert-error').exists()).toBe(true);
        expect(wrapper.find('.auth__alert-error').text().trim()).toBe(initState.ui.msgError);


    });
    

    
});
