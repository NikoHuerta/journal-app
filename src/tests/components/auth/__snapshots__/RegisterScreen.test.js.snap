// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <RegisterScreen /> Debe de mostrarse correctamente 1`] = `
<form
  className="animate__animated animate__fadeIn animate__faster"
  onSubmit={[Function]}
>
  <h3
    className="auth__title"
  >
    Register
  </h3>
  <input
    autoComplete="off"
    className="auth__input"
    name="name"
    onChange={[Function]}
    placeholder="Name"
    type="text"
    value="Nicolas Test"
  />
  <input
    autoComplete="off"
    className="auth__input"
    name="email"
    onChange={[Function]}
    placeholder="Email"
    type="text"
    value="nicolas@huerta.cl"
  />
  <input
    className="auth__input"
    name="password"
    onChange={[Function]}
    placeholder="Password"
    type="password"
    value="123456"
  />
  <input
    className="auth__input"
    name="password2"
    onChange={[Function]}
    placeholder="Confirm Password"
    type="password"
    value="123456"
  />
  <button
    className="btn btn-primary btn-block mb-5"
    type="submit"
  >
    Register
  </button>
  <a
    className="link"
    href="/auth/login"
    onClick={[Function]}
  >
    Already Registered?
  </a>
</form>
`;
